exercises/diffie-hellman/canonical-data.json

```diff
@@ -0,0 +1,67 @@
+{
+  "exercise": "diffie-hellman",
+  "version": "1.0.0",
+  "comments": [
+    "Optional tests to consider:",
+    "* Validation of parameters:",
+    "  Validate that `p`, `g` are valid.",
+    "  Validate that keys given as inputs are valid.",
+    "  Resources that show what happens if parameters are not validated:",
+    "  http://cryptopals.com/sets/5/challenges/34",
+    "  http://cryptopals.com/sets/5/challenges/35",
+    "* Large numbers:",
+    "  Although the calculations fundamentally do not require large numbers,",
+    "  this is a reasonable real-world use for them",
+    "  and it may be instructive to have an exercise on their use.",
+    "  Consult tracks with this exercise (such as the Go track) for possible inputs to use."
+  ],
+  "cases": [
+    {
+      "description": "private key is in range 1 .. p",
+      "property": "privateKeyIsInRange",
+      "input": {},
+      "expected": {}
+    },
+    {
+      "description": "private key is random",
+      "property": "privateKeyIsRandom",
+      "input": {},
+      "expected": {}
+    },
+    {
+      "description": "can calculate public key using private key",
+      "property": "publicKey",
+      "input": {
+        "p": 23,
+        "g": 5,
+        "privateKey": 6
+      },
+      "expected": 8
+    },
+    {
+      "description": "can calculate secret using other party's public key",
+      "property": "secret",
+      "input": {
+        "p": 23,
+        "theirPublicKey": 19,
```

Comment:
it will not be possible to convince me that I should just say `publicKey` and `privateKey` here because I want to make it clear that the public key and private key belong to different people. However, you could convince me of something like `alicePublicKey` and `bobPrivateKey`.

I considered `receiverPublicKey` and `senderPrivateKey` but I thought both parties may play both roles, so I discarded that

Reply:
The next test case uses `alicePublicKey`, so I think it would make sense to have a common terminology.

Reply:
> The next test case uses alicePublicKey, so I think it would make sense to have a common terminology.

I really wanted to emphasise that only I should know my private key. Therefore, I chose `myPrivateKey`. To me, this was not emphasised enough if I simply use `alicePrivateKey`. However, maybe that can be made obvious in an alternate way.

For example, I can rename the keys to `alicePrivateKey` and `bobPublicKey` and have the description be "Alice can calculate a secret using Bob's public key".

Okay. If I use `alicePrivateKey` and `bobPublicKey`, I will insist that I change the description in that way (or any equivalent way to convey the message that only Alice has the power to calculate the secret in this way).

I prefer `myPrivateKey` at a strength of 1/10 so you must tell me if you prefer `alicePrivateKey` and `bobPublicKey` with any strength > 1/10. The scale is subjective, so anything that the person speaking feels comfortable calling > 1/10 is acceptable.